docs/contributing/backend.md

```diff
@@ -0,0 +1,183 @@
+
+## Testing
+
+The Coder backend includes a rich suite of unit and end-to-end tests. A variety of helper utilities are used throughout the codebase to make testing easier, more consistent, and closer to real behavior.
+
+### [clitest](https://github.com/coder/coder/tree/main/cli/clitest)
+
+* Spawns an in-memory `serpent.Command` instance for unit testing
+* Configures an authorized `codersdk` client
+* Once a `serpent.Invocation` is created, tests can execute commands as if invoked by a real user
+
+### [ptytest](https://github.com/coder/coder/tree/main/pty/ptytest)
+
+* `ptytest` attaches to a `serpent.Invocation` and simulates TTY input/output
+* `pty` provides matchers and "write" operations for interacting with pseudo-terminals
+
+### [coderdtest](https://github.com/coder/coder/tree/main/coderd/coderdtest)
+
+* Provides shortcuts to spin up an in-memory `coderd` instance
+* Can start an embedded provisioner daemon
+* Supports multi-user testing via `CreateFirstUser` and `CreateAnotherUser`
+* Includes "busy wait" helpers like `AwaitTemplateVersionJobCompleted`
+* [oidctest](https://github.com/coder/coder/tree/main/coderd/coderdtest/oidctest) can start a fake OIDC provider
+
+### [testutil](https://github.com/coder/coder/tree/main/testutil)
+
+* General-purpose testing utilities, including:
+  * [chan.go](https://github.com/coder/coder/blob/main/testutil/chan.go): helpers for sending/receiving objects from channels (`TrySend`, `RequireReceive`, etc.)
+  * [duration.go](https://github.com/coder/coder/blob/main/testutil/duration.go): set timeouts for test execution
+  * [eventually.go](https://github.com/coder/coder/blob/main/testutil/eventually.go): repeatedly poll for a condition using a ticker
+  * [port.go](https://github.com/coder/coder/blob/main/testutil/port.go): select a free random port
+  * [prometheus.go](https://github.com/coder/coder/blob/main/testutil/prometheus.go): validate Prometheus metrics with expected values
+  * [pty.go](https://github.com/coder/coder/blob/main/testutil/pty.go): read output from a terminal until a condition is met
+
+### [dbtestutil](https://github.com/coder/coder/tree/main/coderd/database/dbtestutil)
+
+* Allows choosing between real and in-memory database backends for tests
+* `WillUsePostgres` is useful for skipping tests in CI environments that don't run Postgres
+
+### [quartz](https://github.com/coder/quartz/tree/main)
+
+* Provides a mockable clock or ticker interface
+* Allows manual time advancement
+* Useful for testing time-sensitive or timeout-related logic
+
+## Quiz
+
+Try to find answers to these questions before jumping into implementation work — having a solid understanding of how Coder works will save you time and help you contribute effectively.
+
+1. When you create a template, what does that do exactly?
+2. When you create a workspace, what exactly happens?
+3. How does the agent get the required information to run?
+4. How are provisioner jobs run?
```

Comment:
This is a cool idea. It'd be nice if there was a document with the answers as well just so you can fact-check your research.

Reply:
Thanks! Just to clarify, this actually came out of a 1:1 with @johnstcn, so the kudos should go to him 🙂 
I thought this could be a helpful resource during onboarding to help people build a mental model of the product early on. 
I’d avoid adding the answers, the idea is to prompt exploration and encourage people to dig into the docs and the codebase on their own to strengthen their understanding. 

While we’re here, maybe we could also add a link to [Deepwiki](https://deepwiki.com/coder/coder), it could be a helpful starting point if diving straight into the source code feels a bit intimidating.

Reply:
Added link to internal onboarding guide.